extensions/local_rate_limit/bucket.cc

```diff
@@ -0,0 +1,142 @@
+#include "extensions/local_rate_limit/bucket.h"
+
+// Key for token bucket shared data.
+constexpr char localRateLimitTokenBucket[] =
+    "wasm_ocal_rate_limit.token_bucket";
+
+// Key for token bucket last updated time.
+constexpr char localRateLimitLastRefilled[] =
+    "wasm_local_rate_limit.last_refilled";
+
+bool getToken() {
+  WasmDataPtr token_bucket_data;
+  uint32_t cas;
+  while (true) {
+    // Get the current token left with cas (compare-and-swap), which will be
+    // used in set call latter.
+    if (WasmResult::Ok !=
+        getSharedData(localRateLimitTokenBucket, &token_bucket_data, &cas)) {
+      return false;
+    }
+    uint64_t token_left =
+        *reinterpret_cast<const uint64_t *>(token_bucket_data->data());
+
+    // If there is no token left, returns false so that request gets 429.
+    if (token_left == 0) {
+      return false;
+    }
+
+    // If there is token left, subtract it by 1, and try set it with cas.
+    // If token bucket set fails because of cas mismatch, which indicates the
+    // bucket is updated by other VMs, retry the whole process.
+    token_left -= 1;
+    auto res = setSharedData(
+        localRateLimitTokenBucket,
+        {reinterpret_cast<const char *>(&token_left), sizeof(token_left)}, cas);
+    if (res == WasmResult::Ok) {
+      // token bucket is updated successfully, returns true and let the request
+      // go through.
+      return true;
+    }
+    if (res == WasmResult::CasMismatch) {
+      continue;
+    }
+    return false;
+  }
+}
+
+void refillToken(uint64_t tokens_per_refill, uint64_t refill_interval_nanosec,
+                 uint64_t max_tokens) {
```

Comment:
Refill could be run in a singleton service , we don’t need so many writers to refill.
Singleton service would also serve as an example.

Reply:
We can leave it as is if too much work.

Reply:
Yeah.. I also considered this. We don't need to make it a singleton, we just need to designate one of the VMs to do token refilling. But I cannot find an easy way to get unique ID for an VM. There are some other ways to work around it, such as getting a timestamp when VM stats and use that as ID. Although it is very unlikely to collide, but it cannot guarantee uniqueness.

Reply:
Create https://github.com/proxy-wasm/proxy-wasm-cpp-host/issues/135 and added a TODO. We can come back on simplify this when that is done.